### PR TITLE
Assembly decode subplan

### PR DESCRIPTION
**Pull request by hensha256**
Should have done this in #212 🤦‍♀️

### CODE REVIEW

File: contracts/modules/uniswap/v3/BytesLib.sol [hunk 1]
```diff
@@ -44,4 +44,12 @@ library BytesLib {
             res.length := calldataload(lengthPtr)
         }
     }
+
+    function toBytesArray(bytes calldata _bytes, uint256 arg) internal pure returns (bytes[] calldata res) {
```
Comment (zhongeric) [inline contracts/modules/uniswap/v3/BytesLib.sol:48]: Maybe dumb q but whats the difference between this and the `toBytes` func above? Just the return type?

Comment (hensha256) [reply contracts/modules/uniswap/v3/BytesLib.sol:48]: Not a stupid question at all - its just the return type indeed. `toBytes`, `toAddressArray` and `toBytesArray` are all the same just different return types lol

Comment (marktoda) [reply contracts/modules/uniswap/v3/BytesLib.sol:48]: possible to have toBytesArray and toAddress just reuse toBytes and force cast? Think worst case you can force cast with assembly walrus haha.. maybe adds a bit of gas for a pointer dup tho

File: contracts/base/Dispatcher.sol [hunk 2]
```diff
@@ -330,7 +330,8 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, Callbacks,
             // 0x20 <= command
         } else {
             if (command == Commands.EXECUTE_SUB_PLAN) {
-                (bytes memory _commands, bytes[] memory _inputs) = abi.decode(inputs, (bytes, bytes[]));
+                bytes calldata _commands = inputs.toBytes(0);
```
Comment (marktoda) [inline contracts/base/Dispatcher.sol:333]: entirely unrelated to the PR but just seeing all the custom calldataloads above for the first time.. sucks we dont have calldata version of abi.decode.. Wonder if there's a nice library or something we can write/use to save from accidentally fatfingering offsets or something

**Review by marktoda — APPROVED** [review]

File: contracts/base/Dispatcher.sol (hunk 2 above)
Comment (marktoda) [reply contracts/base/Dispatcher.sol:333]: ooooh maybe soon tm? https://github.com/ethereum/solidity/issues/13518

Comment (hensha256) [reply contracts/base/Dispatcher.sol:333]: haha yeah the PR that added these decodes was made by the person who made that solidity issue :)

Comment (hensha256) [reply contracts/base/Dispatcher.sol:333]: fingers crossed it happens soon!

**Review by NoahZinsmeister — APPROVED** [review]
nothing to add beyond existing comments!

File: contracts/base/Dispatcher.sol [hunk 3]
```diff
@@ -330,7 +330,8 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, Callbacks,
...
+                bytes[] memory _inputs = inputs.toBytesArray(1);
                 (success, output) =
                     (address(this)).call(abi.encodeWithSelector(Dispatcher.execute.selector, _commands, _inputs));
```
[the remark this answers is not in the document]
Comment (hensha256) [reply contracts/base/Dispatcher.sol:336]: discussed on slack but putting this here for visibility: with delegatecall you could not as easily have a reentrancy lock that allows reentrancy from the contract itself. With `call` you just check `msg.sender==address(this)` but with `delegatecall` you'd have to SLOAD `lockedBy==msg.sender` and it wouldnt be _as_ tight of a lock

[the remark this answers is not in the document]
Comment (shuhuiluo) [reply contracts/base/Dispatcher.sol:336]: I think `_inputs` can be `calldata` as well?

[the remark this answers is not in the document]
Comment (hensha256) [reply contracts/base/Dispatcher.sol:336]: @shuhuiluo youre completely right - great spot. Thanks!

**Pull request merged**